crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_comprehension.rs

```diff
@@ -30,10 +29,23 @@ use crate::rules::flake8_comprehensions::fixes;
 /// set(iterable)
 /// ```
 ///
+/// ## Known problems
+///
+/// When the dictionary key is a tuple, e.g.:
```

Comment:
Hm yeah — I thought the example was suggesting that `for x, y in {}` differed in behavior depending on the type of key. It seems like we just shouldn't raise this violation for dictionaries in general i.e. where `d1` is a dictionary and `.items()` is not called?

```python
d1 = {1: 3, 2: 5}
d2 = {x: y for x, y in d1}
```

Reply:
Yeah, I think that's correct. Possibly we could try to detect if the type of the binding is a dictionary using something like https://github.com/astral-sh/ruff/blob/020f4d4a54eec15b53a5fd94409600f1ad0a63aa/crates/ruff_python_semantic/src/analyze/typing.rs#L752? But I think it's not fully accurate without generalised type inference, and possibly more expensive. Not sure, WDYT?

Reply:
I think we can do something reasonable with that. I'll look into it, but not with a high priority.

Reply:
I guess this only applies to keys that can be unpacked because otherwise that code is a runtime error. Regardless, it seems like we shouldn't depend on that here and should fix the false positive as I described.

Reply:
If we're in a situation where we can statically detect that the keys can't be unpacked but the user appears to be trying to unpack them anyway in a comprehension, then we haven't really got any clue what the user wanted, so I don't think we should emit a violation for that situation either tbh. I'd just not emit the violation ever if we can see that they're iterating over a binding that we know is a dictionary, and they haven't called `.items()` on the dictionary before iterating over it.